[url_launcher] Removed reference to rootViewController during initialization

vitor-gyant opened the pull request:
## Description

The current plugin implementation stores a reference to the rootViewController during the initialisation which breaks scenarios like app2app. This PR does not store any variable to rootViewController but search through the VC hierarchy to get the top most VC in order to present the new vc.

## Related Issues

https://github.com/flutter/flutter/issues/33259

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

cyanglaz commented on packages/url_launcher/ios/Classes/UrlLauncherPlugin.m:
```diff
@@ -165,4 +151,23 @@ - (void)closeWebViewWithResult:(FlutterResult)result API_AVAILABLE(ios(9.0)) {
+  return [self topViewController:[UIApplication sharedApplication].keyWindow.rootViewController];
+}
+
+- (UIViewController *)topViewController:(UIViewController *)rootViewController
```
Add comments on what this method trying to do.
And maybe rename it as `topViewControllerFromViewController:`?

cyanglaz commented on packages/url_launcher/ios/Classes/UrlLauncherPlugin.m:
```diff
@@ -76,18 +70,10 @@ + (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
   UIViewController *viewController =
```
can remove this line

cyanglaz commented on packages/url_launcher/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+
+## 5.1.3
+
+* Fixed iOS app2app scenario.
```
nits:
Always launch url from the top most UIViewController in iOS.

vitor-gyant commented:
@cyanglaz Let me know if there is anything pending to merge the PR on my side. thanks!

cyanglaz commented on packages/url_launcher/ios/Classes/UrlLauncherPlugin.m:
```diff
@@ -165,4 +149,35 @@ - (void)closeWebViewWithResult:(FlutterResult)result API_AVAILABLE(ios(9.0)) {
+ *
+ * @return The top most view controller.
+ */
+- (UIViewController *)topViewControllerFromViewController:(UIViewController *)rootViewController
```
nits: here the name `rootViewController` doesn't make sense anymore since it is a recursive method. Maybe just `viewController`?

vitor-gyant commented:
@cyanglaz I have fixed code format and changed the parameter name. I'm not able to understand why build_all_plugins fails. It seems that name is missing from the pubspec but I didn't remove it. Could you help me on this ? Thanks

vitor-gyant commented:
@cyanglaz Sorry for pushing this but it would be great to have this merged. I'm happy to fix the build_all_plugins issue if you can help me figure out what is the underlying problem. Thanks.

mklim commented:
Sorry, there's been another breakage since then in flutter/flutter that's currently turning CI red now. It will be green again once flutter/flutter#40282 lands.
